src/app/pages/user/dialogs/base/user-details-form/user-details-form.ts

```diff
@@ -0,0 +1,104 @@
+import { Component, OnInit, EventEmitter, Input, Output, OnDestroy } from '@angular/core';
+import { FormBuilder, FormControl, FormGroup, Validators } from '@angular/forms';
+import { User, UserType } from '../../../../../models/user.model';
+import { UserService } from '../../../services/user.service';
+import { ClassService } from '../../../../class/services/class.graphql.service';
+import { Class } from '../../../../../models/class.model';
+
+export type UserTypeKey = keyof typeof UserType;
+
+@Component({
+  selector: 'app-user-details-form',
+  templateUrl: './user-details-form.html',
+  styleUrls: ['./user-details-form.scss'],
+})
+export class UserDetailsFormComponent implements OnInit, OnDestroy {
+  form: FormGroup;
+  roles: string[];
+  userRoleEnum = UserType;
+  classes: Class[];
+  currentRole: UserTypeKey;
+  formControl: FormControl;
+  EmailFormControl: FormControl;
+  selectUserType: FormControl;
+  selectGrade: FormControl;
+  userNameFormControl: FormControl;
+  data: User;
+
+  @Input()
+  options = { submitButtonLabel: 'הוסף' };
+
+  @Input('data')
+  set _data(value: User) {
+    this.data = value;
+    this.currentRole = (value.role as any) as UserTypeKey;
+  }
+
+  @Output()
+  formSubmit = new EventEmitter<User>();
+  @Output()
+  cancel = new EventEmitter<void>();
+
+  constructor(private formBuilder: FormBuilder, private classService: ClassService, public userService: UserService) {
+    this.roles = Object.keys(this.userRoleEnum);
+  }
+
+  ngOnInit(): void {
+    this.attachClasses();
+    this.createFormControls();
+    this.createForm();
+  }
+
+  onUserTypeChange(userType: UserTypeKey): void {
+    this.currentRole = userType;
+    if (this.userRoleEnum[userType] === this.userRoleEnum.PRINCIPLE && this.data.class) {
+      this.data.class = undefined;
+    }
+  }
+  getClassDisabled() {
+    return this.userRoleEnum[this.currentRole] !== this.userRoleEnum.TEACHER;
+  }
+
+  attachClasses() {
+    this.classService.getAllClasses().subscribe((classes) => {
+      this.classes = [...classes];
+    });
+  }
+
+  close() {
+    this.cancel.emit();
+  }
+
+  submitForm() {
+    this.formSubmit.emit(this.form.value);
+  }
+
+  compareClass(c1: Class, c2: Class): boolean {
+    return c1 && c2 ? c1._id === c2._id : c1 === c2;
```

Comment:
why you explain this function?

Reply:
Hi  @dimkinv,
This function is used for showing the right initial value on class select box

Reply:
Yes. What I'm trying to understand is why so many assertions needed? Does this mean that the data is not consistent?